cockroachdb/templates/job.init.yaml

```diff
@@ -119,29 +121,140 @@ spec:
+                    --host={{ template "cockroachdb.fullname" . }}-0.{{ template "cockroachdb.fullname" . -}}
+                            :{{ .Values.service.ports.grpc.internal.port | int64 }} \
+                    --execute="
+                      {{- range $clusterSetting, $clusterSettingValue := .Values.init.provisioning.clusterSettings }}
+                        SET CLUSTER SETTING {{ $clusterSetting }} = '${{ $clusterSetting | replace "." "_" }}_CLUSTER_SETTING';
+                      {{- end }}
+
+                      {{- range $user := .Values.init.provisioning.users }}
+                        CREATE USER IF NOT EXISTS {{ $user.name }} WITH
+                        {{- if $user.password }}
+                          PASSWORD '${{ $user.name }}_PASSWORD'
+                        {{- else }}
+                          PASSWORD null
+                        {{- end }}
+                          {{ join " " $user.options }}
+                        ;
+                      {{- end }}
+
+                      {{- range $database := .Values.init.provisioning.databases }}
+                        CREATE DATABASE IF NOT EXISTS {{ $database.name }}
+                          {{- if $database.options }}
+                            {{ join " " $database.options }}
+                          {{- end }}
+                        ;
+
+                      {{- range $owner := $database.owners }}
+                        GRANT ALL ON DATABASE {{ $database.name }} TO {{ $owner }};
+                      {{- end }}
+
+                      {{- if $database.backup }}
+                        CREATE SCHEDULE {{ $database.name }}_scheduled_backup
+                          FOR BACKUP DATABASE {{ $database.name }} INTO '{{ $database.backup.into }}'
+
+                        {{- if $database.backup.options }}
+                          WITH {{ join "," $database.backup.options }}
+                        {{- end }}
+                          RECURRING '{{ $database.backup.recurring }}'
+                        {{- if $database.backup.fullBackup }}
+                          FULL BACKUP '{{ $database.backup.fullBackup }}'
+                        {{- else }}
+                          FULL BACKUP ALWAYS
+                        {{- end }}
+
+                        {{- if and $database.backup.schedule $database.backup.schedule.options }}
+                          WITH SCHEDULE OPTIONS {{ join "," $database.backup.schedule.options }}
+                        {{- end }}
+                        ;
+                      {{- end }}
+                      {{- end }}
+
+                        SELECT 'Provisioning completed successfully';
+                    "
+                  &>/dev/null;
+
+                  local exitCode="$?";
+
+                  if [[ "$exitCode" == "0" ]]
+                    then break;
+                  fi
+
+                  sleep 5;
+                done
+              }
+
+              provisionCluster;
```

Comment:
Can this be wrapped in a timeout so that it eventually breaks the loop?

Reply:
Hmm... Can you please provide an example?

Reply:
What happens if `/cockroach/cockroach sql` exits with a non-zero value, when does the init container exit the while loop?

I was thinking of the [timeout command](https://man7.org/linux/man-pages/man1/timeout.1.html) so something like `timeout 30 provisionCluster;` is there something else that will cause the init container to stop after a certain amount of time?

Reply:
If sql command exit with non-zero code you will get infinite loop with 5 seconds interval. You should get logs and try to find root cause of non completed job. Same behaviour as before, now init job is doing the same

What will happen if sql script is valid, but not ran to completion in 30 seconds? I think that completed job without any guarantee of applied sql is unclear and mostly harmful

Also timeout configuration can be very tricky

Reply:
And the sql script is idempotent? The 30 seconds was arbitrary but you're right that half applied sql statements are more likely to put the DB into a bad state. If the script is continuously failing though won't the DB be in a bad state already potentially? @keith-mcclellan are you ok with this workflow?

Reply:
Yes, all except backups are idempontent https://github.com/cockroachdb/cockroach/issues/57892

Reply:
Ok thanks for the clarification